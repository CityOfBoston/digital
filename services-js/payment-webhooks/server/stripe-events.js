// @flow
/* eslint no-console: 0 */

import type { NodeStripe, Event, Charge } from 'stripe';
import type INovah, { INovahFactory } from './services/iNovah';

export type Dependencies = {|
  inovahFactory: INovahFactory,
|};

async function processChargeSucceeded(
  stripe: NodeStripe,
  inovah: INovah,
  charge: Charge
): Promise<void> {
  // The Charge object only has the gross amount charged to the customer, so we
  // need to look up the associated transaction to find the net amount that will
  // be deposited into our account.
  const balanceTransaction = await stripe.balance.retrieveTransaction(
    charge.balance_transaction
  );

  // Stripe works in cents, iNovah in floating-point dollars.
  await inovah.addTransaction(balanceTransaction.net / 100);
}

export async function processStripeEvent(
  stripe: NodeStripe,
  inovah: INovah,
  event: Event
): Promise<void> {
  console.log('STRIPE WEBHOOK: ', JSON.stringify(event));

  switch (event.type) {
    case 'charge.succeeded':
      await processChargeSucceeded(stripe, inovah, event.data.object);
      break;
    default:
      break;
  }
}
