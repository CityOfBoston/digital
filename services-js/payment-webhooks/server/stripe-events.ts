/* eslint no-console: 0 */

import Stripe, { webhooks, charges } from 'stripe';
import INovah from './services/INovah';

import Rollbar from 'rollbar';

type Dependencies = {
  inovah: INovah;
  rollbar: Rollbar;
  stripe: Stripe;
};

/**
 * Called when a charge is created or finally captured to update iNovah with the
 * transaction information, and then update the charge with the iNovah reference
 * info.
 */
async function processChargeSucceeded(
  { rollbar, stripe, inovah }: Dependencies,
  charge: charges.ICharge
): Promise<void> {
  // Birth cert charges aren't captured at first, we wait until the Registry
  // department fulfills the order to do that. Treasury doesn't want to hear
  // about the money until itâ€™s captured, so we exit out here without sending
  // it to iNovah.
  if (!charge.captured) {
    console.log(`Skipping charge ${charge.id}: not yet captured`);
    return;
  }

  // The Charge object only has the gross amount charged to the customer, so we
  // need to look up the associated transaction to find the net amount that will
  // be deposited into our account.
  const latestCharge = await stripe.charges.retrieve(charge.id, {
    expand: ['balance_transaction'],
  });

  if (typeof latestCharge.balance_transaction === 'string') {
    throw new Error('balance_transaction not expanded');
  }

  if (
    latestCharge.metadata['inovah.transactionId'] &&
    !process.env.SKIP_IDEMPOTENCY_CHECKS
  ) {
    console.log(
      `Charge ${charge.id} already has an iNovah transaction ID saved`
    );
    return;
  }

  const balanceTransaction = latestCharge.balance_transaction;
  const orderType = charge.metadata['order.orderType'] || 'DC';

  const source = charge.source;

  if (source.object !== 'card') {
    throw new Error(`Unexpected source type: ${source.object}`);
  }

  // Stripe works in cents, iNovah in floating-point dollars.
  const amountInDollars = balanceTransaction.net / 100;

  const {
    transactionId,
    transactionNum,
    batchId,
    batchNum,
  } = await inovah.addTransaction(
    charge.metadata['order.orderId'] || 'unknown',
    charge.id,
    balanceTransaction.id,
    orderType,
    {
      amountInDollars,
      quantity: parseInt(charge.metadata['order.quantity'] || '0', 10),
      unitPriceInDollars:
        parseInt(charge.metadata['order.unitPrice'] || '0', 10) / 100,
    },
    {
      cardholderName: source.name || '',
      billingAddress1: source.address_line1 || '',
      billingAddress2: source.address_line2 || '',
      billingCity: source.address_city || '',
      billingState: source.address_state || '',
      billingZip: source.address_zip,
    }
  );

  if (process.env.NODE_ENV !== 'test') {
    console.log(`Added charge ${charge.id} to iNovah:`, {
      transactionId,
      batchId,
      amountInDollars,
    });
  }

  try {
    await stripe.charges.update(charge.id, {
      metadata: {
        'inovah.transactionId': transactionId,
        'inovah.transactionNum': transactionNum,
        'inovah.batchId': batchId,
        'inovah.batchNum': batchNum,
      },
    });
  } catch (e) {
    // Don't fail things if we aren't able to back-update Stripe
    rollbar.error(e);
  }
}

export async function processStripeEvent(
  deps: Dependencies,
  webhookSecret: string | undefined,
  webhookSignature: string,
  body: string
): Promise<void> {
  const event: webhooks.StripeWebhookEvent<any> = webhookSecret
    ? deps.stripe.webhooks.constructEvent(body, webhookSignature, webhookSecret)
    : JSON.parse(body);

  if (process.env['NODE_ENV'] !== 'test') {
    console.log(
      'STRIPE WEBHOOK: ',
      JSON.stringify({ ...event, data: 'REDACTED' })
    );
  }

  switch (event.type) {
    case 'charge.succeeded':
      await processChargeSucceeded(deps, event.data.object as charges.ICharge);
      break;
    case 'charge.captured':
      await processChargeSucceeded(deps, event.data.object as charges.ICharge);
      break;
    default:
      break;
  }
}
