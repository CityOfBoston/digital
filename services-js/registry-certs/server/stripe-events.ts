/* eslint no-console: 0 */
import RegistryDb from './services/RegistryDb';
import Emails from './services/Emails';
import Stripe, { charges, webhooks } from 'stripe';

import { loadOrder } from './graphql/death-certificates';

import {
  FIXED_CC_SERVICE_FEE,
  PERCENTAGE_CC_SERVICE_FEE,
  SERVICE_FEE_URI,
} from '../lib/costs';

interface Dependencies {
  emails: Emails;
  stripe: Stripe;
  registryDb: RegistryDb;
}

// We handle Stripe’s "charge.succeeded" event to complete the order process:
//   - Mark the order as "paid" in the backend so that fulfillment can complete
//   - Send a receipt email
//
// We do this from Stripe’s webhook to take advantage of their reliability
// features. If the charge succeeded, THE ORDER MUST BE PLACED for the customer.
// If we tried to do this inline in the mutation after the charge API call,
// there’s a chance that the database write would fail, or the email wouldn’t
// send.
//
// If those things happen in this case, we’ll send a 500 back to Stripe and it
// will retry. If it keeps failing, Stripe will notify us of the issue.
//
// Note that there is a non-zero chance that Stripe will call this twice for the
// same charge, meaning we will send 2 emails to the customer about the order.
// That’s a better choice than not sending an email at all or causing delays in
// fulfillment, so we can live with it.
export async function processChargeSucceeded(
  { emails, registryDb }: Dependencies,
  charge: charges.ICharge
) {
  const {
    metadata: {
      'webapp.name': webappName,
      'webapp.nodeEnv': webappNodeEnv,
      'order.orderKey': orderKey,
      'order.orderId': orderId,
    },
  } = charge;

  // We're going to get every Stripe event, so make sure that we’e only handling
  // our own. Also, we do the NODE_ENV check so that charges made in dev (which
  // still go to Test Stripe) don’t cause staging to send an email.
  if (
    webappName !== 'registry-certs' ||
    webappNodeEnv !== process.env.NODE_ENV
  ) {
    return;
  }

  // We slightly-hackily rely on the DB order -> JS object code from the GraphQL
  // side of things. This could be more principled.
  const order = await loadOrder(registryDb, orderId);

  if (!order) {
    throw new Error(`Order ${orderId} not found in the database`);
  }

  // By marking the payment as successful we can tell Registry they can proceed
  // with fulfillment. This method is idempotent.
  await registryDb.addPayment(
    parseInt(orderKey, 10),
    // Unix epoch seconds -> milliseconds
    new Date(charge.created * 1000),
    charge.id,
    // Per Rich, this should be the total amount charged by Stripe, including
    // their fee, not just the subtotal we receive.
    charge.amount / 100
  );

  // Sending the email is not idempotent, so we put it last and in the vast
  // majority of cases it will run once.
  await emails.sendReceiptEmail(order.contactName, order.contactEmail, {
    orderId: order.id,
    orderDate: order.date,
    shippingName: order.shippingName,
    shippingCompanyName: order.shippingCompanyName,
    shippingAddress1: order.shippingAddress1,
    shippingAddress2: order.shippingAddress2,
    shippingCity: order.shippingCity,
    shippingState: order.shippingState,
    shippingZip: order.shippingZip,
    subtotal: order.subtotal,
    serviceFee: order.serviceFee,
    total: order.total,
    fixedFee: FIXED_CC_SERVICE_FEE,
    percentageFee: PERCENTAGE_CC_SERVICE_FEE,
    serviceFeeUri: SERVICE_FEE_URI,
    // loadOrder comes from the GraphQL side of things first and foremost, so it
    // returns an async function for the certificate so that the extra DB lookup
    // (needed for certificate name) doesn’t have to happen if it’s not
    // requested by the query. Since we do need the name for the receipt, we
    // have to trigger and then resolve the promises.
    items: await Promise.all(
      order.items.map(async ({ id, quantity, cost, certificate }) => ({
        id,
        quantity,
        cost,
        name: nameFromCertificate(await certificate()),
      }))
    ),
  });
}

function nameFromCertificate(cert) {
  if (cert) {
    return `${cert.firstName} ${cert.lastName}`;
  } else {
    // typically should not happen, but we need to guard anyway
    return 'UNKNOWN CERTIFICATE';
  }
}

export async function processStripeEvent(
  deps: Dependencies,
  webhookSecret: string | undefined,
  webhookSignature: string,
  body: string
): Promise<void> {
  const event: webhooks.StripeWebhookEvent<any> = webhookSecret
    ? deps.stripe.webhooks.constructEvent(body, webhookSignature, webhookSecret)
    : JSON.parse(body);

  if (process.env['NODE_ENV'] !== 'test') {
    console.log(
      'STRIPE WEBHOOK: ',
      JSON.stringify({ ...event, data: 'REDACTED' })
    );
  }

  switch (event.type) {
    case 'charge.succeeded':
      await processChargeSucceeded(deps, event.data.object as charges.ICharge);
      break;
    default:
      break;
  }
}
