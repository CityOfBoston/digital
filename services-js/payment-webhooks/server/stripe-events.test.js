// @flow

import { processStripeEvent } from './stripe-events';

import CHARGE_SUCCEEDED from '../fixtures/stripe/charge-succeeded';
import BALANCE_TRANSACTION from '../fixtures/stripe/balance-transaction';

describe('processStripeEvent', () => {
  let stripe: any;
  let inovah: any;

  beforeEach(() => {
    stripe = {
      balance: {
        retrieveTransaction: jest.fn(),
      },
    };

    inovah = {
      addTransaction: jest.fn(),
    };
  });

  describe('charge.succeeded', () => {
    beforeEach(() => {});
    test('it sends charge to iNovah', async () => {
      stripe.balance.retrieveTransaction.mockReturnValue(
        Promise.resolve(BALANCE_TRANSACTION)
      );

      await processStripeEvent(stripe, inovah, CHARGE_SUCCEEDED);
      expect(inovah.addTransaction).toHaveBeenCalledWith(14.0);
    });
    test('it rejects if Stripe fails', async () => {
      stripe.balance.retrieveTransaction.mockReturnValue(
        Promise.reject(new Error('Stripe backend error!'))
      );

      expect(
        processStripeEvent(stripe, inovah, CHARGE_SUCCEEDED)
      ).rejects.toMatchSnapshot();
    });

    test('it rejects if iNovah fails', async () => {
      stripe.balance.retrieveTransaction.mockReturnValue(
        Promise.resolve(BALANCE_TRANSACTION)
      );

      inovah.addTransaction.mockReturnValue(
        Promise.reject(new Error('iNovah backend error!'))
      );

      expect(
        processStripeEvent(stripe, inovah, CHARGE_SUCCEEDED)
      ).rejects.toMatchSnapshot();
    });
  });
});
