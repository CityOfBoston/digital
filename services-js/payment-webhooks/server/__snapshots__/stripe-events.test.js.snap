// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processStripeEvent charge.succeeded it rejects if Stripe fails 1`] = `[Error: Stripe backend error!]`;

exports[`processStripeEvent charge.succeeded it rejects if iNovah fails 1`] = `[Error: iNovah backend error!]`;
